fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,546 @@
+                        "Trying to activate node at index {}",
+                        ingest_summary_node_mapping.node_index
+                    );
+                    match self.ingest_clients[ingest_summary_node_mapping.node_index].activate() {
+                        Ok(_) => {
+                            log::info!(self.logger, "Successfully activated ingest client.");
+                            return Ok(());
+                        }
+                        Err(_) => {
+                            // TODO: See if we can make an id field for ingest client. This
+                            // will allow us to better identify ingest clients.
+                            log::warn!(self.logger, "Could not activate idle ingest client.");
+                        }
+                    }
+                }
+            }
+        }
+
+        Err(OverseerError::ActivateNode("Could not activate a node because no node has an ingress key that matches the outstanding key.".to_string()))
+    }
+
+    fn report_lost_ingress_key(
+        &self,
+        inactive_outstanding_key: CompressedRistrettoPublic,
+    ) -> Result<(), OverseerError> {
+        let result = retry_with_index(Fixed::from_millis(200), |current_try| {
+            if current_try as u8 > Self::NUMBER_OF_TRIES {
+                return OperationResult::Err(OverseerError::ReportLostKey(format!(
+                    "Did not suceed in reporting lost ingress key {} within {} tries.",
+                    inactive_outstanding_key,
+                    Self::NUMBER_OF_TRIES,
+                )));
+            }
+
+            match self
+                .recovery_db
+                .report_lost_ingress_key(inactive_outstanding_key)
+            {
+                Ok(_) => {
+                    let success_message = format!(
+                        "The following key  was successfully reported as lost: {}",
+                        inactive_outstanding_key
+                    );
+                    OperationResult::Ok(success_message)
+                }
+                Err(_) => {
+                    let number_of_remaining_tries = Self::NUMBER_OF_TRIES - current_try as u8;
+                    let error_message = format!("The following key  was not successfully reported as lost: {}. Will try {} more times", inactive_outstanding_key, number_of_remaining_tries);
+                    OperationResult::Retry(OverseerError::ReportLostKey(error_message))
+                }
+            }
+        });
+
+        match result {
+            Ok(success_message) => {
+                log::info!(self.logger, "{}", success_message);
+                Ok(())
+            }
+            // TODO: Add alerting that says we need to manually report these keys
+            // as lost.
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    fn set_new_key_on_a_node(&self) -> Result<usize, OverseerError> {
+        for (i, ingest_client) in self.ingest_clients.iter().enumerate() {
```

Comment:
I think we could assert that we're only calling this on idle nodes as a precaution.

Reply:
I need to think about putting in an assert for idle state. I agree that it might be good for precaution's sake, but this method is only called when the main method, `perform_automatic_failover`, has determined that every node for which it received a response, is idle. 

I guess the question is: Is it possible for one of these idle nodes to become active in between when `perform_automatic_failover` determines that it's idle and when this method is called? In a perfect world, the answer is no. I'll think more on it to see if there any cases that might violate this assumption. If there are, then the assert makes perfect sense. 

@garbageslam any thoughts on this?

Reply:
I think the server already checks this basically -- it won't allow you to call new_keys if the ingest server is in the active state. I think that's the right layer to guard against this.

That being said, it probably can't hurt to put an assert here though. But I don't think it's essential.

Reply:
Yeah, agreed that the server layer is the right layer to check this. I'll leave as is for now, but if we decide to change in the future, will do.